### PR TITLE
Ensure that kinds are checked by meet_block_field_simple

### PR DESCRIPTION
**Pull request by mshinwell**
The fast path in `simplify_immutable_block_load`, which uses `T.meet_block_field_simple`, is not currently checking the kind of any `Simple` found in the corresponding field against the required kind.  This can cause `Simple`s of the wrong kind to be returned with corresponding wrong equations added to the environment.

### CODE REVIEW

File: middle_end/flambda2/simplify/lifting/lifted_constant.ml [hunk 1]
```diff
@@ -292,8 +292,8 @@ let apply_projection t proj =
     let meet_shortcut =
       match Symbol_projection.projection proj with
       | Block_load { index } ->
-        T.meet_block_field_simple typing_env ~min_name_mode:Name_mode.normal ty
-          index
+        T.meet_block_field_simple typing_env ~min_name_mode:Name_mode.normal
+          ~field_kind:Flambda_kind.value ty index
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/simplify/lifting/lifted_constant.ml]: I've now propagated the kind on all symbol projections.

**Comment by mshinwell** [comment]
> Looks good apart from the symbol projection, where I'm not sure. I would also have expected the prover to check the kind directly on the `int_indexed_product`, but looking at the code it seems simpler to just check the kind of the resulting type. Maybe we could leave a comment explaining that ?

Comments added.

File: middle_end/flambda2/simplify/expr_builder.ml [hunk 2]
```diff
@@ -520,18 +520,36 @@ let create_let_symbols uacc lifted_constant ~body =
...
+                match Flambda_kind.With_subkind.kind kind with
+                | Value ->
+                  let field_kind : P.Block_access_field_kind.t =
+                    if Flambda_kind.With_subkind.equal kind
+                         Flambda_kind.With_subkind.tagged_immediate
+                    then Immediate
+                    else Any_value
+                  in
+                  Values
+                    { tag = Known Tag.Scannable.zero;
```
Comment (lthls) [inline middle_end/flambda2/simplify/expr_builder.ml:537]: This is not directly linked to this PR, but I don't understand why `Known Tag.Scannable.zero` is correct here.

File: middle_end/flambda2/types/provers.ml [hunk 3]
```diff
@@ -771,19 +776,24 @@ let meet_project_function_slot_simple env ~min_name_mode t function_slot :
   | Naked_nativeint _ | Rec_info _ | Region _ ->
     wrong_kind "Value" t
 
-let meet_project_value_slot_simple env ~min_name_mode t env_var :
-    Simple.t meet_shortcut =
+let meet_project_value_slot_simple env ~min_name_mode ~value_slot_kind t env_var
```
Comment (lthls) [inline middle_end/flambda2/types/provers.ml:779]: Here `env_var` already carries a kind, so the new check is still welcome but I don't think we need the new parameter.

File: middle_end/flambda2/simplify/expr_builder.ml (hunk 2 above)
Comment (mshinwell) [reply middle_end/flambda2/simplify/expr_builder.ml]: Fixed.  I think `Unknown` here is ok (`To_cmm` ignores this in particular, I think).

**Review by lthls — APPROVED** [review]

**Pull request merged**